llvm/lib/Transforms/Vectorize/VPlan.h

```diff
@@ -1166,6 +1166,30 @@ class VPIRInstruction : public VPRecipeBase {
   void extractLastLaneOfFirstOperand(VPBuilder &Builder);
 };
 
+/// Helper type to provide functions to access incoming values and blocks for
+/// phi-like recipes. RecipeTy must be a sub-class of VPRecipeBase.
+template <typename RecipeTy> class VPPhiAccessors {
+  /// Return a VPRecipeBase* to the current object.
+  const VPRecipeBase *getAsRecipe() const {
+    return static_cast<const RecipeTy *>(this);
```

Comment:
Is the templating really needed, can `getAsRecipe()` simply down cast `this` to `VPRecipeBase`?

Reply:
Yep it seems so, problem is that w/o the template there's no inheritance relation here, and static_casts are rejected (Curiously Recurring Template Pattern). For some reason, we need to cast exactly to the type.

But the template parameter may cause problems in the future, so I updated `getAsRecipe` to be pure virtual to be implemented by the derived classes.  The single implementation could also be used for other trait classes in the future. WDYT?

Reply:
The use case enabled by getting rid of the template argument is supporting `dyn_cast` from `VPRecipeBase` -> `VPPhiAccessors`, as used in https://github.com/llvm/llvm-project/pull/124838 ( https://github.com/llvm/llvm-project/pull/124838/files#diff-a69094b5fcfce6b2bf9e957e2ac7011e5492e81c885129506c90874375e621fbR210) by implementing `CastInfo<VPPhiAccessors, const VPRecipeBase *>`

Reply:
Hmm, would having getAsRecipe() do a dyn_cast to VPRecipeBase be ok? Possibly followed by an assert...

Reply:
I don't think we could  `dyn_cast` from `VPPhiAccessors` to `VPRecipeBase` because there's no directy relationship between them (and the `this` pointer for the `VPPhiAccessors` (sub-)object may not be the same as the VPRecipeBase (base-) object. The other way works, because we can down-cast to the concrete types inheriting from `VPPhiAccessors`).